### PR TITLE
[SPARK-12340][SQL] Fix overstep the bounds of Int in SparkPlan.executeTake

### PR DESCRIPTION
**Pull request by QiangCai**
If partsScanned = 1, numPartsToTry = Int.MaxValue and totalParts = 200, the result of math.min(partsScanned + numPartsToTry, totalParts) will be Int.MinValue (-2147483648).

Modifies partsScanned to partsScanned.toLong and  chang result of math.min to Int.

https://issues.apache.org/jira/browse/SPARK-12340

### CODE REVIEW

**Comment by 3ourroom** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <[spark] [SPARK-12340][SQL] Fix overstep the bounds of Int in SparkPlan.executeTake (#10310)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by 3ourroom** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-12340][SQL] Fix overstep the bounds of Int in SparkPlan.executeTake (#10310)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 1]
```diff
@@ -206,7 +206,7 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
       numPartsToTry = math.max(0, numPartsToTry)  // guard against negative num of partitions
 
       val left = n - buf.size
-      val p = partsScanned until math.min(partsScanned + numPartsToTry, totalParts)
+      val p = partsScanned until math.min(partsScanned.toLong + numPartsToTry, totalParts).toInt
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: Sounds good to me.

**Comment by QiangCai** [comment]
Tthe number of  partitions which has scanned is  the size of seq p, not numPartsToTry.

**Comment by srowen** [comment]
Yes I think that also makes sense, in the case where we hit the limit of `totalParts`.

**Comment by rxin** [comment]
Can you also change take in RDD?

**Comment by QiangCai** [comment]
Yes. I will change all.

**Comment by QiangCai** [comment]
Yes. I will check all at first.

**Comment by srowen** [comment]
@QiangCai I think that technically `var partsScanned = 0` in `take` is a problem since it's incremented by `numPartsToTry` and could overflow causing `partsScanned < totalParts` to always be true. This is a very rare case where `numPartsToTry` is large and `totalParts` is close to the max integer, and `n` is very large.

**Comment by QiangCai** [comment]
@srowen I have modified all the code, and try to keep them to be the same.

At first, the var numPartsToTry and partsScanned  have been set to Long.
Secondly,  because  var p should be Seq[Int], so i have added to.Int  in the until statement

**Comment by srowen** [comment]
@QiangCai thanks that looks good, but this needs a rebase now.

**Comment by QiangCai** [comment]
@srowen I maybe have made a mistake in git bash. Can I new another pull request to resolved this problem.

**Comment by srowen** [comment]
I think you just need to rebase from master and force-push the result, but do what you need to.

**Comment by QiangCai** [comment]
@srowen I have new another pull request. This pull request will be closed.
another pull request https://github.com/apache/spark/pull/10487

**Pull request closed without merging**